lib/iris/fileformats/netcdf.py

```diff
@@ -1103,20 +1107,23 @@ def __init__(self, filename, netcdf_format):
         #: A dictionary, mapping formula terms to owner cf variable name
         self._formula_terms_cache = {}
         #: NetCDF dataset
-        try:
-            self._dataset = netCDF4.Dataset(
-                filename, mode="w", format=netcdf_format
-            )
-        except RuntimeError:
-            dir_name = os.path.dirname(filename)
-            if not os.path.isdir(dir_name):
-                msg = "No such file or directory: {}".format(dir_name)
-                raise IOError(msg)
-            if not os.access(dir_name, os.R_OK | os.W_OK):
-                msg = "Permission denied: {}".format(filename)
-                raise IOError(msg)
-            else:
-                raise
+        if hasattr(filename, "createVariable"):
```

Comment:
Is it worth considering a runtime protocol to define, in code, what you mean by "supporting the netCF4.Dataset api"?

https://docs.python.org/3/library/typing.html#typing.runtime_checkable. Added in Python 3.8.